File: MB-Whatsapp/observability/telemetry-flags.ts
import type { ObservabilityFlags } from "./types";

export const DEFAULT_OBSERVABILITY_FLAGS: ObservabilityFlags = {
  enabled: false,
  shadowMode: true,
  structuredLoggingEnabled: false,
  traceEngineEnabled: false,
  providerMetricsEnabled: false,
  confidenceMetricsEnabled: false,
  reviewMetricsEnabled: false,
  multimodalMetricsEnabled: false,
  externalExportEnabled: false,
  phiAllowed: false,
};

function readBoolean(value: string | undefined, fallback: boolean): boolean {
  if (value === undefined) return fallback;
  return value.trim().toLowerCase() === "true";
}

export function loadObservabilityFlags(env: NodeJS.ProcessEnv = process.env): ObservabilityFlags {
  return {
    enabled: readBoolean(env.OBSERVABILITY_ENABLED, DEFAULT_OBSERVABILITY_FLAGS.enabled),
    shadowMode: readBoolean(env.OBSERVABILITY_SHADOW_MODE, DEFAULT_OBSERVABILITY_FLAGS.shadowMode),
    structuredLoggingEnabled: readBoolean(env.OBSERVABILITY_STRUCTURED_LOGGING_ENABLED, DEFAULT_OBSERVABILITY_FLAGS.structuredLoggingEnabled),
    traceEngineEnabled: readBoolean(env.OBSERVABILITY_TRACE_ENGINE_ENABLED, DEFAULT_OBSERVABILITY_FLAGS.traceEngineEnabled),
    providerMetricsEnabled: readBoolean(env.OBSERVABILITY_PROVIDER_METRICS_ENABLED, DEFAULT_OBSERVABILITY_FLAGS.providerMetricsEnabled),
    confidenceMetricsEnabled: readBoolean(env.OBSERVABILITY_CONFIDENCE_METRICS_ENABLED, DEFAULT_OBSERVABILITY_FLAGS.confidenceMetricsEnabled),
    reviewMetricsEnabled: readBoolean(env.OBSERVABILITY_REVIEW_METRICS_ENABLED, DEFAULT_OBSERVABILITY_FLAGS.reviewMetricsEnabled),
    multimodalMetricsEnabled: readBoolean(env.OBSERVABILITY_MULTIMODAL_METRICS_ENABLED, DEFAULT_OBSERVABILITY_FLAGS.multimodalMetricsEnabled),
    externalExportEnabled: readBoolean(env.OBSERVABILITY_EXTERNAL_EXPORT_ENABLED, DEFAULT_OBSERVABILITY_FLAGS.externalExportEnabled),
    phiAllowed: readBoolean(env.OBSERVABILITY_PHI_ALLOWED, DEFAULT_OBSERVABILITY_FLAGS.phiAllowed),
  };
}

File: api/app/runtime_support/observability/telemetry_flags.py
from __future__ import annotations

import os
from collections.abc import Mapping

from .types import ObservabilityFlags


def _read_bool(env: Mapping[str, str], name: str, fallback: bool) -> bool:
    value = env.get(name)
    if value is None:
        return fallback
    return value.strip().lower() == "true"


def load_observability_flags(env: Mapping[str, str] | None = None) -> ObservabilityFlags:
    source = env or os.environ
    return ObservabilityFlags(
        enabled=_read_bool(source, "OBSERVABILITY_ENABLED", False),
        shadow_mode=_read_bool(source, "OBSERVABILITY_SHADOW_MODE", True),
        structured_logging_enabled=_read_bool(source, "OBSERVABILITY_STRUCTURED_LOGGING_ENABLED", False),
        trace_engine_enabled=_read_bool(source, "OBSERVABILITY_TRACE_ENGINE_ENABLED", False),
        provider_metrics_enabled=_read_bool(source, "OBSERVABILITY_PROVIDER_METRICS_ENABLED", False),
        confidence_metrics_enabled=_read_bool(source, "OBSERVABILITY_CONFIDENCE_METRICS_ENABLED", False),
        review_metrics_enabled=_read_bool(source, "OBSERVABILITY_REVIEW_METRICS_ENABLED", False),
        multimodal_metrics_enabled=_read_bool(source, "OBSERVABILITY_MULTIMODAL_METRICS_ENABLED", False),
        external_export_enabled=_read_bool(source, "OBSERVABILITY_EXTERNAL_EXPORT_ENABLED", False),
        phi_allowed=_read_bool(source, "OBSERVABILITY_PHI_ALLOWED", False),
    )
